### PR TITLE
AAP-29426: Automate two reports and save output in csv files

### PR DESCRIPTION
**Pull request by manstis**
Jira Issue: https://issues.redhat.com/browse/AAP-29426

## Description
This PR automates production of the "Trial User reports".

It adds a Django admin command that is triggered by a k8s `CronJob` [in](https://github.com/ansible/ansible-wisdom-ops/pull/950) `ansible-wisdom-ops`.

The command extracts the data and builds CSV response. It is sent to a configurable "postman".

I have implemented "postmen" for Slack Web Hooks, Slack WebAPI and GDrive.

Slack Web Hooks do not support attachments; so the report is inline in the Slack message.

image

## Testing
- Create a namespace somewhere.
- Deploy `Secret`'s[*] `wisdom-service`, `wisdom-service-deployment` and `wca-awssm` to the namespace
- Deploy `ConfigMap`[*] `wisdom-service` to the namespace.
- Deploy https://github.com/ansible/ansible-wisdom-ops/pull/950 to the namespace; with the following changes:-
  - Change `schedule` to `*/5 * * * *` (every 5 minutes for testing)
  - Additional environment variables:
```
env:
  - name: ANSIBLE_AI_ONE_CLICK_REPORTS_POSTMAN
    value: "slack-webhook"
  - name: ANSIBLE_AI_ONE_CLICK_REPORTS_CONFIG
    value: '{"slack-webhook-url": <I can advise>}'
```              
  - Change `image` to `image: quay.io/ansible/ansible-ai-connect-service:pr-1260-ddc4c663ab1e05943f7fe62e800d844130db2761`

You should notice posts to the configured Slack channel every minutes.

[*] The content of these resources can be copied from `stage2-west`, `wisdom-service-staging` namespace.

### Scenarios tested
As above.

## Production deployment
- [ ] This code change is ready for production on its own
- [x] This code change requires the following considerations before going to production:
https://github.com/ansible/ansible-wisdom-ops/pull/950

### CODE REVIEW

File: ansible_ai_connect/users/management/commands/generate_users_trials_reports.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        )
+
+    def handle(self, dry_run, plan_id, created_after, created_before, auto_range, *args, **options):
+        plan_id = int(plan_id) if plan_id else None
+        created_after = created_after if created_after else None
+        created_before = created_before if created_before else None
+
+        if auto_range:
+            created_before = timezone.now()
+            created_after = created_before - relativedelta(days=1)
```
Comment (jameswnl) [inline ansible_ai_connect/users/management/commands/generate_users_trials_reports.py:76]: IIRC, cron job doesn't guarantee to run exactly on the specified time due to scheduling/load etc.

So using this, there is possibly we will miss a small window in-between 2 daily runs.

File: ansible_ai_connect/users/management/commands/generate_users_trials_reports.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+        if dry_run:
+            self.stdout.write("Reports not sent due to --dry-run parameter.")
```
Comment (jameswnl) [inline ansible_ai_connect/users/management/commands/generate_users_trials_reports.py:79]: do you want to still continue except skip the upload?

File: ansible_ai_connect/main/settings/base.py [hunk 3]
```diff
@@ -582,3 +585,38 @@
...
+ANSIBLE_AI_ONE_CLICK_REPORTS_SLACK_TOKEN: str = (
+    os.getenv("ANSIBLE_AI_ONE_CLICK_REPORTS_SLACK_TOKEN") or ""
+)
+ANSIBLE_AI_ONE_CLICK_REPORTS_SLACK_CHANNEL_ID: str = (
+    os.getenv("ANSIBLE_AI_ONE_CLICK_REPORTS_SLACK_CHANNEL_ID") or ""
+)
+ANSIBLE_AI_ONE_CLICK_REPORTS_SLACK_WEBHOOK_URL: str = (
+    os.getenv("ANSIBLE_AI_ONE_CLICK_REPORTS_SLACK_WEBHOOK_URL") or ""
+)
+ANSIBLE_AI_ONE_CLICK_REPORTS_GOOGLE_DRIVE_FOLDER_NAME: str = (
```
Comment (jameswnl) [inline ansible_ai_connect/main/settings/base.py:604]: can we have just one env var expecting a json string to include all these stuff?

File: ansible_ai_connect/users/management/commands/generate_users_trials_reports.py (hunk 1 above)
Comment (manstis) [reply ansible_ai_connect/users/management/commands/generate_users_trials_reports.py:76]: `--auto-range` now defaults the time elements to "midnight".

For example, `created_before=2024-12-31 00:00:00`, `created_after=2024-12-30 00:00:00`.

File: ansible_ai_connect/users/management/commands/generate_users_trials_reports.py (hunk 2 above)
Comment (manstis) [reply ansible_ai_connect/users/management/commands/generate_users_trials_reports.py:79]: I now use the `StdoutPostman` for `--dry-run` which prints the reports to the logger.

File: ansible_ai_connect/main/settings/base.py (hunk 3 above)
Comment (manstis) [reply ansible_ai_connect/main/settings/base.py]: OK. Done.

**Comment by manstis** [comment]
@jameswnl This is ready for another review.

I tested successfully in `stage2-west` with the `SlackWebhookPostman` i.e. I see the reports in my `#manstis-test` channel.

I still await ownership of the GDrive application to complete testing.

**Review by jameswnl — APPROVED** [review]

**Pull request merged**